### PR TITLE
perf(sessions): coalesce each flush batch into one store append per notification run

### PR DESCRIPTION
**Pull request by KarloAldrete**
## Problem

The 16ms flush introduced in #3063 batches *processing*, but not *store writes*: `flushSessionEvents` still hands its buffered events to `handleSessionEvent` one at a time, so a 15-event burst does 15 `setState` calls through the session store. Each write is O(transcript) — immer produces a new session map and every store subscriber gets notified — so during heavy streaming the store write path alone costs, measured on a real 30k-event transcript, **~248ms of main-thread time per streamed second** (V8, see below). It scales linearly with transcript length, which is why long-running tasks get progressively more janky while streaming: components subscribed to the session (`useSessionForTask` callers) get re-render-notified ~900 times/second.

## Changes

`handleSessionEvents` (new) applies one flush batch per task: runs of plain notifications — agent message chunks and tool-call updates, the bulk of a turn — are coalesced into a **single `appendEvents` call**, followed by one `updatePromptStateFromEvents` for the run (it already takes arrays).

- Events carrying a JSON-RPC `id` (user-prompt echoes, stop-reason responses) still go through `handleSessionEvent` **individually, unchanged**, so turn-lifecycle semantics (optimistic replacement, `currentPromptId` matching, queue draining, prompt-complete notifications) are untouched.
- Global order is preserved: the pending run is flushed before each id-carrying event is handled.
- The notification side effects (config/usage/adapter/compaction updates) split out into `applyNotificationEffects`, shared verbatim by both paths.

Complements #3132 (which adds `useSessionSelector` to narrow the *read* side); this PR narrows the *write* side. No overlap in files.

## How did you test this?

- New unit tests in `sessionEventBatching.test.ts`: a 3-chunk burst produces exactly **one** `appendEvents` call in order; an id-carrying echo splits the run but preserves global transcript order; a stop-reason response batched together with chunks still fires `notifyPromptComplete`. (The harness `updateSession` mock now replaces the session object like the real store, instead of mutating in place — the stop-reason check depends on pre-update snapshots.)
- Full suites: `@posthog/core` 1,972 passed, `@posthog/ui` sessions 317 passed; core typecheck and Biome clean.
- Measured with the included `scripts/measure-append-hotpath.ts` (imports the real store module; run with `tsx` on V8): 60 flushes × 15 events over a seeded transcript, median of 7 runs:

| Transcript | per-event (main) | coalesced (this PR) | |
| --- | --- | --- | --- |
| 10,000 events | 20.9 ms/s | 2.9 ms/s | ~7x |
| 30,000 events | 248.1 ms/s | 25.9 ms/s | **~10x** |

Store subscriber notifications during streaming drop from ~900/s to ~60/s.

Honest caveats: numbers are store-write CPU measured in isolation (Node/V8), not end-to-end frame timings in the app; and a handwritten (immer-free) `appendEvents` was also tried and measured **within noise** of the immer version on the real module, so it was deliberately left out to keep the diff minimal.

## Automatic notifications

- [ ] Publish to changelog?
- [ ] Alert Sales and Marketing teams?

---
*Created with [PostHog Code](https://posthog.com/code?ref=pr)*

### CODE REVIEW

File: packages/core/src/sessions/sessionEventBatching.test.ts [hunk 1]
```diff
@@ -149,6 +183,60 @@ describe("streamed event batching", () => {
...
+    h.emit(promptEcho(7, "user steer-less prompt"));
+    h.emit(chunk("c"));
+    vi.advanceTimersByTime(FLUSH_MS);
+
+    // Two coalesced appends around the individually-handled echo.
+    expect(h.appendEvents).toHaveBeenCalledTimes(2);
+    expect(h.appendEvents.mock.calls[0][1].map(chunkText)).toEqual(["a", "b"]);
+    expect(h.appendEvents.mock.calls[1][1].map(chunkText)).toEqual(["c"]);
+    expect(h.events().map(chunkText)).toEqual(["a", "b", "c"]);
+  });
```
[the remark this answers is not in the document]
Comment (KarloAldrete) [reply packages/core/src/sessions/sessionEventBatching.test.ts]: Done in 4abb864 — the two cases are now an `it.each` table of (emitted events → expected append groups) pairs, plus two new edge cases: consecutive id-carrying events (no empty appends) and a single-notification batch.

File: scripts/measure-append-hotpath.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
+// Mide el hot path real de appendEvents (módulo real, sin réplicas):
+// 1 segundo de streaming pesado = 60 flushes x 15 eventos sobre un transcript
+// de N eventos. Correr en main y en la rama para comparar.
+//   bun scripts/measure-append-hotpath.ts
```
[the remark this answers is not in the document]
Comment (KarloAldrete) [reply scripts/measure-append-hotpath.ts]: Done in 4abb864 — comments and output labels are now in English (matching the PR description's table wording).

**Comment by KarloAldrete** [comment]
Rebased on latest main. Two things worth noting:

- **Still applies on top of #3063**: main's flush path still does one store write per event (`flushSessionEvents` → per-event `handleSessionEvent` → `appendEvents(taskRunId, [acpMsg])`, sessionService.ts:1539-1546). #3063 made each write cheaper (autofreeze off, per-event freeze); this PR reduces the write *count* per notification run from N to 1 — complementary, not overlapping.
- **#3166's turn-duration test integrated**: its `promptEcho`/`stopResponse` helpers are unified into the parameterised harness (`promptEcho(id, text, ts)`), and its "keeps the turn duration" case passes through the coalesced path unchanged — id-carrying responses still go through `handleSessionEvent` one-by-one, so `turnStartedAtTs` semantics are untouched.

All 269 session tests green locally (`vitest run src/sessions/`).

**Comment by charlesvien** [comment]
Closing stale PRs ahead of Friday's monorepo migration. If this is a mistake, rebase and reopen for review.
